<a href="https://colab.research.google.com/github/abdullahattiq786/abdullah-flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/abdullahattiq786/abdullah-flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

# Week 1 — Research Question and Lane Selection

## Chosen Lane
Lane 2 — Refresh / Content Opportunity Scoring


## 1. Search Question

**Which pages should be reviewed first for refresh, expansion, protection, pruning, or monitoring?**

The goal is to determine whether observable content and search-performance signals can be used to prioritize pages for human review.

The system is intended to support prioritization rather than guarantee that a particular SEO action will improve a page's performance.

## 2. Unit of Analysis

**One row represents one content/page observation.**

The decision will be made at the page/content level. Each page can be evaluated using observable signals such as impressions, clicks, sessions, CTR, average position, content age, freshness, engagement, and trend information.

The purpose is to rank individual pages according to how strongly the available evidence suggests that they deserve review.


## 3. Output

The intended output is a **ranked review queue of pages**.

For each page, the eventual output should include:

- a priority/opportunity score;
- a suggested action such as refresh, expansion, protection, pruning, or monitoring;
- reason codes explaining why the page received its score;
- a confidence label.

The output should help a human reviewer understand both **which pages to review first** and **why they were prioritized**.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

## 4. Action

A FlyRank reviewer could use the ranked queue to decide which pages to investigate first.

Depending on the signals associated with a page, the reviewer could consider actions such as refreshing content, expanding content, protecting a strong page, pruning a weak page, or monitoring the page.

The model or scoring system would support human decision-making rather than automatically applying an SEO change.

## 5. Cost of a Wrong Recommendation

A false positive could cause a reviewer to spend limited time investigating a page that does not represent a meaningful opportunity. This could delay work on pages that deserve more attention.

A false negative could also be costly because a page with a genuine opportunity or risk might be ranked too low and receive less attention.

Because reviewer capacity is limited, the quality of the highest-ranked recommendations is especially important.

Therefore, evaluation should focus on ranking quality, including metrics such as Precision@K and average precision, rather than relying only on overall classification accuracy.


## 6. Why Data / ML Can Help

A content inventory can contain a large number of pages, making it difficult for a human reviewer to compare every page consistently.

The starter dataset contains observable signals about content and search performance. These signals can be combined to identify pages that may deserve review.

A transparent rule-based score can provide a simple baseline. Machine learning can then be used to test whether combinations of multiple signals improve the ranking of useful review candidates.

The purpose of using data and ML is therefore not to replace human judgment, but to reduce the search space and direct limited reviewer attention toward pages that appear most worth investigating.

In [15]:
import os

print("Current directory:")
print(os.getcwd())

print("\nFiles/folders here:")
print(os.listdir())

Current directory:
/content

Files/folders here:
['.config', 'abdullah-flyrank-ml-internship', 'sample_data']


In [16]:
 !git clone https://github.com/abdullahattiq786/abdullah-flyrank-ml-internship.git


fatal: destination path 'abdullah-flyrank-ml-internship' already exists and is not an empty directory.


In [17]:
import os

print(os.listdir("/content"))

['.config', 'abdullah-flyrank-ml-internship', 'sample_data']


In [18]:
import os

repo_path = "/content/abdullah-flyrank-ml-internship"

print(os.listdir(repo_path))

['submission', 'DATA_USE.md', 'notebooks', '.github', 'outputs', 'GUIDE.md', 'requirements.txt', 'CLAUDE.md', 'docs', 'scripts', 'README.md', 'skills', 'data', 'SETUP.md', 'AGENTS.md', '.gitignore', 'LICENSE', 'work', '.git']


In [19]:
print(os.listdir(repo_path + "/data"))

['raw']


In [20]:
print(os.listdir(repo_path + "/data/raw"))

['content_refresh_anonymized.csv']


In [21]:
import pandas as pd

DATA_PATH = "/content/abdullah-flyrank-ml-internship/data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Dataset shape:", df.shape)

df.head()

Dataset shape: (30000, 44)


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [22]:
print("Number of rows:", len(df))
print("Number of columns:", len(df.columns))

print("\nColumns:")
print(df.columns.tolist())

Number of rows: 30000
Number of columns: 44

Columns:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [23]:
num_pages = len(df)

print(f"Number of page observations: {num_pages:,}")

Number of page observations: 30,000


In [24]:
high_visibility_pages = (df["impressions_90d"] >= 500).sum()

print(
    f"Pages with at least 500 impressions in the last 90 days: "
    f"{high_visibility_pages:,}"
)

Pages with at least 500 impressions in the last 90 days: 16,726


In [25]:
declining_pages = (df["trend_direction"] == "down").sum()

print(
    f"Pages currently marked with a downward trend: "
    f"{declining_pages:,}"
)

Pages currently marked with a downward trend: 16,262


In [26]:
high_visibility_pct = high_visibility_pages / len(df) * 100
declining_pct = declining_pages / len(df) * 100

print(f"Pages with >=500 impressions: {high_visibility_pct:.2f}%")
print(f"Pages with downward trend: {declining_pct:.2f}%")

Pages with >=500 impressions: 55.75%
Pages with downward trend: 54.21%


## 7. Evidence From the Starter Dataset

The starter dataset contains **30,000 page observations**, providing a meaningful number of page-level observations for investigating a prioritization problem.

Of these, **16,726 pages have at least 500 impressions in the last 90 days**. This indicates that a measurable portion of the dataset has meaningful search visibility and can potentially be considered when prioritizing review candidates.

Additionally, **16,262 pages are currently marked with a downward trend**. This provides a measurable group of pages showing a current performance signal that may warrant further investigation.

These numbers make the Refresh / Content Opportunity Scoring lane worth exploring because the dataset has both sufficient page-level scale and observable variation in page visibility and current performance signals.


## 8. Why This Lane Is Worth the Next 7 Weeks

This lane is worth pursuing because it addresses a concrete decision: deciding which pages a reviewer should investigate first when review capacity is limited.

The starter dataset provides page-level observations and multiple observable signals that can potentially be combined into a useful prioritization system.

Over the next seven weeks, I would investigate whether a simple and interpretable baseline can be improved using additional signals and machine-learning models.

I would also improve the problem definition by developing a stronger future-looking target where appropriate, validating the ranking without leakage, examining the highest-ranked pages manually, and producing reason codes so that recommendations remain understandable to reviewers.

The final goal is not to guarantee that a recommended page will improve after an SEO change. Instead, the goal is to determine whether data-driven prioritization can make page review more efficient and useful.

## 9. Initial Limitations

The starter dataset is useful for establishing the problem, but the current declining label is a proxy based on the current window rather than a true future outcome.

Therefore, I will not interpret a downward-trend label as proof that a page will continue declining or that refreshing it will cause recovery.

Future work should define prediction and feature windows carefully, avoid using future information as features, consider minimum-volume requirements, and validate whether the ranking generalizes beyond the starter dataset.